tasks/rh-sign-image-cosign/rh-sign-image-cosign.yaml

```diff
@@ -73,28 +82,68 @@ spec:
             echo "No valid file was provided as signRegistryAccessPath."
             exit 1
         fi
+        PUBLIC_KEY_FILE=$(mktemp)
+        echo -n "$PUBLIC_KEY" > "$PUBLIC_KEY_FILE"
+        RUNNING_JOBS="\j" # Bash parameter for number of jobs currently running
 
         run_cosign () { # Expected arguments are [digest_reference, tag_reference]
-            # Upload transparency log when rekor url is specified
-            if [ -v REKOR_URL ]; then
-                COSIGN_COMMON_ARGS="-y --rekor-url=$REKOR_URL --key $SIGN_KEY"
-            else
-                COSIGN_COMMON_ARGS="--tlog-upload=false --key $SIGN_KEY"
-            fi
-            echo "Signing manifest $1 ($2)"
             attempt=0
+            backoff1=2
+            backoff2=3
             until [ "$attempt" -gt "$(params.retries)" ] ; do # 3 retries by default
-                cosign -t 3m0s sign\
-                ${COSIGN_COMMON_ARGS}\
-                --sign-container-identity "$2"\
-                "$1" && break
+                cosign "$@" && break
+                sleep $backoff2
+
+                # Fibbonaci backoff
+                old_backoff1=$backoff1
+                backoff1=$backoff2
+                backoff2=$((old_backoff1 + backoff2))
+
                 attempt=$((attempt+1))
             done
             if [ "$attempt" -gt "$(params.retries)" ] ; then
               echo "Max retries exceeded."
               exit 1
             fi
         }
+        function check_existing_signatures() {
+          local identity=$1
+          local reference=$2
+          local digest=$3
+          if [ -v REKOR_URL ]; then
+              COSIGN_REKOR_ARGS="--rekor-url=$REKOR_URL"
+          else
+              COSIGN_REKOR_ARGS="--insecure-ignore-tlog=true"
+          fi
+          verify_output=$(run_cosign verify "$COSIGN_REKOR_ARGS" "$reference")
+          found_signatures=$(echo "$verify_output" | jq -j '[.[]|select(.critical.image."docker-manifest-digest"'
+          '| contains("'"$digest"'"))|'
+          'select(.critical.identity."docker-reference"| contains("'"$identity"'"))]|length')
+          echo "$found_signatures"
+        }
+        function check_and_sign() {
+          local identity=$1
+          local reference=$2
+          local tag=$3
+          local digest=$4
+          found_signatures=$(check_existing_signatures "$identity" "$reference:$tag" "$digest")
+          if [ -z "$found_signatures" ]; then
+            found_signatures=0
+          fi
+          echo "FOUND SIGNATURES: $found_signatures"
```

Comment:
Maybe print a line that will contain all the details at once and if it exists and will be skipped or not? Because of multiple runs running in parallel, it will be hard to figure out what happened from the log. So if we have something like "Signature for $identity with digest $digest is already signed, skipping" in the log, I think that would help - I'm not sure what are the best identifying params to print, it's up to you.

Reply:
How about creating special echo function which will print job id as prefix before every line? I was also thinking about way to identify which log belongs to which request